### DISCUSSION TITLE
v1.8.* 讨论区

### DISCUSSION THREAD

**Discussion by ckken** [empjs/emp#88] [Ideas]
## 1 emp-config 优化 [已完成]
+  json exports 方式 代替 function exports  让配置更加语义化，更加直观的调用、需要向下兼容当前写法，逐步迁移
```javascript
declare interface EMPConfig {
  /** webpack & webpack chain config method */
  webpack?: (o: EmpConfigI & EmpConfigIBase) => WebpackConfigI | Promise<WebpackConfigI>
  webpackChain?: (config: webpackChain, o: EmpConfigI) => void | Promise<any>
  /**
   * compile replace babel use swc or esbuild
   * 拓展 swc、build 替换 babel进行构建
   */
  compile?: Array<any>
  /**
   * react vue svetle and more
   * 拓展 js 框架支持
   */
  framework?: Array<any>
  /**
   * module federation config
   * module federation 设置
   */
  moduleFederation?: ModuleFederationType
  /**
   * global assets path
   * when output.publicPath=auto & use module federation,need setting this option
   * 当 output.publicPath=auto时 自定义静态文件路径
   */
  moduleGenerator?: ModuleGeneratorType
  /**
   * style command of emp
   * 命令行拓展
   */
  commander?: (program: Command) => void
  /**
   * before dev env=development server
   * emp dev 启动前执行
   */
  beforeDev?: (config: EmpConfigI) => void | Promise<any>
  /**
   * before build env=production server
   * emp build 启动前执行
   */
  beforeBuild?: (config: EmpConfigI) => void | Promise<any>
}
``````
+ 增加对 [emp-config.js](https://github.com/efoxTeam/emp/blob/main/projects/demo1/emp-config.js) 对 command 的扩展

## 2 插件化接入EMP  [已完成]
+ 业务、公司内部服务等定制模板命令行接入
+ 自动化触发构建接入

## 3 引用外部基站可视化  [重新评估]
> 考虑放到 v1.9 重点实现
+ 分析项目所引用的所有基站
+ 可视化显示关联基站
+ 一件同步外部基站类型文件

**Comment by morance** [comment]
webpack => configureWebpack 内部使用webpack merge， 这是vue-cli使用的key，目的应该也是一大块的自定义方便写入。
可以把一下常用的配置提升譬如entry之类的。

**Reply by ckken** [reply]
类似于 https://github.com/efoxTeam/emp/issues/85 我比较担心后续有冲突的问题 如果 是按照 webpack 的原来方式来可能更加规范些

**Reply by morance** [reply]
个人觉得如果json配置具有一些key是自己管理的比较合理，毕竟你可以通过json schema做提示等补充操作

**Comment by flyyuan** [comment]
我准备尝试实现：
    2 插件化接入EMP
    业务、公司内部服务等定制模板命令行接入 -->已完成，给 init 命令增加 emp init -t <remote-template-url>

**Comment by ckken** [comment]
## 暂时定义
```javascript 
import {Configuration, container} from 'webpack/types'
import Config from 'webpack-chain/types'
type MFOptions = ConstructorParameters<typeof container.ModuleFederationPlugin>[0]
declare interface EMPConfig {
  /** webpack config */
  webpack?: (wp: Configuration, empEnv: string, webpackEnv: 'development' | 'production') => Configuration
  /** webpack chain config */
  webpackChain?: (wpc: Config, empEnv: string, webpackEnv: 'development' | 'production') => void
  /** compile replace babel use swc or esbuild */
  compile?: Array<any>
  /** react vue svetle and more */
  framework?: Array<any>
  /** module federation config */
  moduleFederation?: MFOptions
}
```

**Reply by ckken** [reply]
![image](https://user-images.githubusercontent.com/3890513/110420800-0c712e00-80d7-11eb-814f-84fefb01cd6a.png)

**Comment by ckken** [comment]
## 优化内容：TODO
+ 验证windows 为什么重复打开新tab问题
+ 减少不必要的引用 精简emp-cli 安装库
+ 考虑增加 dev 模式下 dist 的生成 与生产保持一致

**Comment by ckken** [comment]
## ts 配置  emp
+  demo https://github.com/efoxTeam/emp/blob/v1.8.0/projects/demo1/emp-config.ts 
+ 配置内容 https://github.com/efoxTeam/emp/blob/v1.8.0/packages/emp-cli/types/emp-config.d.ts

**Comment by ckken** [comment]
`yarn dev --ts` 支持本地访问类型文件

**Comment by flyyuan** [comment]
# emp 插件 from branch v1.8.0-register-plugin

TODO:
- [x] register 改为 registerCommand
- [x] 全局插件名改为 emp-plugin-*

## 1. 局部插件，仅用于当前项目:
例如：
>emp-extra.js
```javascript
register({
  name: 'hello',
  options: [{name: '-i, --item <item>', description: 'flavour of pizza'}],
  exec: ({item}) => {
    console.log(`hell ${item}`)
  },
})
```
![image](https://user-images.githubusercontent.com/19996552/113371489-16661400-9399-11eb-9404-9806c1670cbb.png)

## 2. 全局插件，包名前缀需要为 `emp-*`, `index.js` 为 emp 全局插件入口。
全局安装  emp 插件，例如：
>/Users/shibaoyuan/.nvm/versions/node/v12.16.2/lib/node_modules/emp-test/index.js
```javascript
register({
  name: 'helloGlobal',
  options: [{name: '-i, --item <item>', description: 'flavour of pizza'}],
  exec: ({item}) => {
    console.log(`hello global ${item}`)
  },
})

```
![image](https://user-images.githubusercontent.com/19996552/113373526-9bebc300-939d-11eb-9d6d-75e1bb7c36bc.png)

**Reply by ckken** [reply]
good 补充一下 如果 非全局指令 不想以 emp- 为前缀 可以考虑使用 https://github.com/efoxTeam/emp/blob/d6abb7000e4c8400cf3e819ba3379cc2e393beaf/projects/esbuild-react-demo/emp-config.ts#L19

**Reply by flyyuan** [reply]
> good 补充一下 如果 非全局指令 不想以 emp- 为前缀 可以考虑使用
> 
> https://github.com/efoxTeam/emp/blob/d6abb7000e4c8400cf3e819ba3379cc2e393beaf/projects/esbuild-react-demo/emp-config.ts#L19

marker

**Comment by ckken** [comment]
避免非ts 项目需要用 类型提示 考虑增加 js 的类型提示函数引入

**Comment by ckken** [comment]
## emp-config.js 增加类型提示的方法
```js
/**
 * @type {import('@efox/emp-cli').EMPConfig}
 */
module.exports = {
  webpack({webpackEnv}) {
    console.log('webpack', webpackEnv)
    return {
      devServer: {
        port: 8001,
      },
    }
  },
  moduleGenerator({webpackEnv}) {
    console.log('moduleGenerator', webpackEnv)
    return webpackEnv === 'development' ? '/' : `http://localhost:8001/`
  },
  async moduleFederation({webpackEnv}) {
    return mf(webpackEnv)
  },
}

```

**Reply by ckken** [reply]
emp-config.ts 将弃置

**Comment by longkehuawei** [comment]
[1/4] Resolving packages...
success Already up-to-date.
Done in 0.73s.
yarn run v1.22.10
$ emp dev
╔════════════════════╤═══════════════╗
║ Emp & Deps         │ Version       ║
╟────────────────────┼───────────────╢
║ @efox/emp-cli      │ 1.8.10        ║
╟────────────────────┼───────────────╢
║ webpack            │ ^5.28.0       ║
╟────────────────────┼───────────────╢
║ webpack-dev-server │ ^4.0.0-beta.1 ║
╟────────────────────┼───────────────╢
║ typescript         │ ^4.1.3        ║
╟────────────────────┼───────────────╢
║ postcss            │ ^8.2.8        ║
╚════════════════════╧═══════════════╝

config { host: 'localhost',
  port: 6007,
  publicPath: 'http://localhost:6007/' }

* Webpack █████████████████████████ setup (3%)
 watch run

(node:19640) UnhandledPromiseRejectionWarning: Error: Compiling RuleSet failed: Unexpected property test in condition (at ruleSet[1].
rules[1].resource.test: resource => {
        currentResource = resource
        return true
      })
    at RuleSetCompiler.error (D:\toB\ui\node_modules\@efox\emp-cli\node_modules\webpack\lib\rules\RuleSetCompiler.js:373:10)
    at RuleSetCompiler.compileCondition (D:\toB\ui\node_modules\@efox\emp-cli\node_modules\webpack\lib\rules\RuleSetCompiler.js:309:1
7)
    at ruleSetCompiler.hooks.rule.tap (D:\toB\ui\node_modules\@efox\emp-cli\node_modules\webpack\lib\rules\BasicMatcherRulePlugin.js:
29:40)
    at Hook.eval [as call] (eval at create (D:\toB\ui\node_modules\@efox\emp-cli\node_modules\tapable\lib\HookCodeFactory.js:19:10),
<anonymous>:17:1)
    at RuleSetCompiler.compileRule (D:\toB\ui\node_modules\@efox\emp-cli\node_modules\webpack\lib\rules\RuleSetCompiler.js:177:19)
    at rules.map (D:\toB\ui\node_modules\@efox\emp-cli\node_modules\webpack\lib\rules\RuleSetCompiler.js:154:9)
    at Array.map (<anonymous>)
    at RuleSetCompiler.compileRules (D:\toB\ui\node_modules\@efox\emp-cli\node_modules\webpack\lib\rules\RuleSetCompiler.js:153:16)
    at RuleSetCompiler.compileRule (D:\toB\ui\node_modules\@efox\emp-cli\node_modules\webpack\lib\rules\RuleSetCompiler.js:184:30)
    at rules.map (D:\toB\ui\node_modules\@efox\emp-cli\node_modules\webpack\lib\rules\RuleSetCompiler.js:154:9)
    at Array.map (<anonymous>)
    at RuleSetCompiler.compileRules (D:\toB\ui\node_modules\@efox\emp-cli\node_modules\webpack\lib\rules\RuleSetCompiler.js:153:16)
    at RuleSetCompiler.compile (D:\toB\ui\node_modules\@efox\emp-cli\node_modules\webpack\lib\rules\RuleSetCompiler.js:68:22)
    at new NormalModuleFactory (D:\toB\ui\node_modules\@efox\emp-cli\node_modules\webpack\lib\NormalModuleFactory.js:238:34)
    at Compiler.createNormalModuleFactory (D:\toB\ui\node_modules\@efox\emp-cli\node_modules\webpack\lib\Compiler.js:1020:31)
    at Compiler.newCompilationParams (D:\toB\ui\node_modules\@efox\emp-cli\node_modules\webpack\lib\Compiler.js:1040:30)
(node:19640) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an asy
nc function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:19640) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are no
t handled will terminate the Node.js process with a non-zero exit code.
<i> [webpack-dev-server] Project is running at:
<i> [webpack-dev-server] Loopback: http://localhost:6007/, http://127.0.0.1:6007/
<i> [webpack-dev-server] Content not from webpack is served from 'D:\toB\ui\public' directory
<i> [webpack-dev-server] 404s will fallback to '/index.html'
Starting server on http://localhost:6007
<i> [webpack-dev-middleware] wait until bundle finished: /


为啥接入报错呀

**Reply by ckken** [reply]
发 issue 附带个代码 codepen 谢谢

**Comment by ckken** [comment]
esbuild 全局安装收到影响 导致 cli有问题 需要 `npm install --unsafe-perm` 解决，考虑去掉 emp-config.ts 使用 emp-config.js (已经支持类型提示)代替

**Reply by longkehuawei** [reply]
https://github.com/lanxingmo/hello-dev  有时间帮我看下是啥原因呀，用了vue脚手架，集成emp，报错了

**Reply by longkehuawei** [reply]
https://github.com/lanxingmo/hello-dev 有时间帮忙看下，vue脚手架，集成emp 报错了
